File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/HexNumberFormat.java
package com.cedarsoft.commons.javafx;

import java.text.FieldPosition;
import java.text.NumberFormat;
import java.text.ParsePosition;

/**
 * formats integers to/from hex string
 */
public class HexNumberFormat extends NumberFormat {
  @Override
  public StringBuffer format(double number, StringBuffer toAppendTo, FieldPosition pos) {
    throw new IllegalArgumentException("doubles not supported.");
  }

  @Override
  public StringBuffer format(long number, StringBuffer toAppendTo, FieldPosition pos) {
    StringBuffer stringBuffer = toAppendTo.append("0x").append(String.format("%08X", number));
    return stringBuffer;
  }

  @Override
  public Number parse(String source, ParsePosition parsePosition) {
    long l = 0;
    if (source.startsWith("0x")) {
      try {
        l = Long.parseLong(source.substring(2), 16);
      }
      catch (NumberFormatException e) {
        e.printStackTrace();
      }
      return l;
    }

    try {
      l = Long.parseLong(source, 16);
    }
    catch (NumberFormatException e) {
      e.printStackTrace();
    }
    return l;
  }
}

File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/HexNumberFormat.kt
package com.cedarsoft.commons.javafx

import java.text.FieldPosition
import java.text.NumberFormat
import java.text.ParsePosition
import java.util.Locale

/**
 * formats integers to/from hex string
 */
class HexNumberFormat : NumberFormat() {
  override fun format(number: Double, toAppendTo: StringBuffer, pos: FieldPosition): StringBuffer {
    throw IllegalArgumentException("doubles not supported.")
  }

  override fun format(number: Long, toAppendTo: StringBuffer, pos: FieldPosition): StringBuffer {
    return toAppendTo.append("0x").append(String.format(Locale.getDefault(), "%08X", number))
  }

  override fun parse(source: String, parsePosition: ParsePosition): Number {
    if (source.startsWith("0x")) {
      try {
        return source.substring(2).toLong(16)
      } catch (e: NumberFormatException) {
        e.printStackTrace()
      }
      return 0
    }

    try {
      return source.toLong(16)
    } catch (e: NumberFormatException) {
      e.printStackTrace()
    }
    return 0
  }
}
